fsspec/implementations/cache_mapper.py

```diff
@@ -30,14 +30,38 @@ def __hash__(self) -> int:
 
 
 class BasenameCacheMapper(AbstractCacheMapper):
-    """Cache mapper that uses the basename of the remote URL.
+    """Cache mapper that uses the basename of the remote URL and a fixed number
+    of directory levels above this.
 
-    Different paths with the same basename will therefore have the same cached
-    basename.
+    The default is zero directory levels, meaning different paths with the same
+    basename will have the same cached basename.
     """
 
+    def __init__(self, directory_levels: int = 0):
+        if directory_levels < 0:
+            raise ValueError(
+                "BasenameCacheMapper requires zero or positive directory_levels"
+            )
+        self.directory_levels = directory_levels
+
+        # Separator for directories when encoded as strings.
+        self._separator = "_^_"
+
     def __call__(self, path: str) -> str:
-        return os.path.basename(path)
+        dirname, basename = os.path.split(path)
+
+        if self.directory_levels > 0:
+            dirs = dirname.split(os.sep)[-self.directory_levels :]
+            dirs.append(basename)
+            basename = self._separator.join(dirs)
+
+        return basename
```

Comment:
The use of `os.sep` here probaby broke the windows test?

Also, I think it might be simpler:

```suggestion
        prefix, *bits = dirname.rsplit(separator, self.directory_labels + 1)
        return self._separator.join(bits)
```

Reply:
Yes, your simpler approach works.